File: Advanced Core/Django_RoR_Symfony--Piscines/Piscine-RoR/2. RoR - 0 - Oob/ex05/Page.rb
#!/usr/bin/env ruby -w
require_relative 'ex03'
require_relative 'ex04'

class Page
  def initialize(elem)
    raise TypeError, "Page must be initialized with an Elem instance." unless elem.is_a?(Elem)
    @elem = elem
  end

  def is_valid?
    @valid_output = []
    res = _is_recursive_valid?(@elem)
    puts @valid_output.join("\n") if res
    puts "FILE IS OK" if res
    res
  end

  def _is_recursive_valid?(elem)
    # Node type check
    allowed_types = [Html, Head, Body, Title, Meta, Img, Table, Th, Tr, Td, 
                     Ul, Ol, Li, H1, H2, P, Div, Span, Hr, Br, Text]
    
    # Check if instance of one of the allowed classes
    is_allowed = allowed_types.any? { |t| elem.is_a?(t) }
    return false unless is_allowed

    # If it's Text, it's valid (leaf node)
    if elem.is_a?(Text)
      # @valid_output << "Currently evaluating a Text :\n-Text -> Must contains a simple string\nText content is OK"
      return true
    end

    # Specific rule checks
    @valid_output << "Currently evaluating a #{elem.class} :"
    
    case elem
    when Html
      @valid_output << "- root element of type 'html'" if elem == @elem
      @valid_output << "- Html -> Must contains a Head AND a Body after it"
      return false if elem.content.count != 2
      return false unless elem.content[0].is_a?(Head) && elem.content[1].is_a?(Body)
      @valid_output << "Head is OK" # Example shows this after head check
    when Head
      @valid_output << "- Head should contain only one Title."
      return false if elem.content.count != 1 || !elem.content[0].is_a?(Title)
    when Body, Div
      @valid_output << "- Body and Div should only contain H1, H2, Div, Table, Ul, Ol, Span or Text."
      body_allowed = [H1, H2, Div, Table, Ul, Ol, Span, Text]
      return false unless elem.content.all? { |c| body_allowed.any? { |t| c.is_a?(t) } }
    when Title, H1, H2, Li, Th, Td
      @valid_output << "- Title, H1, H2, Li, Th, Td should only contain one Text."
      return false if elem.content.count != 1 || !elem.content[0].is_a?(Text)
    when P
      @valid_output << "- P should only contain Text elements."
      return false unless elem.content.all? { |c| c.is_a?(Text) }
    when Span
      @valid_output << "- Span should only contain Text or P elements."
      return false unless elem.content.all? { |c| c.is_a?(Text) || c.is_a?(P) }
    when Ul, Ol
      @valid_output << "- Ul and Ol must contain at least one Li and only Li elements."
      return false if elem.content.count < 1 || !elem.content.all? { |c| c.is_a?(Li) }
    when Tr
      @valid_output << "- Tr must contain at least one Th or Td and only Th or Td elements. Th and Td should be mutually exclusive."
      return false if elem.content.count < 1
      first_type = elem.content[0].class
      return false unless [Th, Td].include?(first_type)
      return false unless elem.content.all? { |c| c.is_a?(first_type) }
    when Table
      @valid_output << "- Table should only contain Tr elements."
      return false unless elem.content.all? { |c| c.is_a?(Tr) }
    when Img
        @valid_output << "- Img: should have a src field and its value should be of type Text."
        return false if elem.opt[:src].nil? || !elem.opt[:src].is_a?(Text)
        @valid_output << "Img content is OK"
    end

    # Check content recursively
    elem.content.each do |child|
        @valid_output << "Evaluating a multiple node" if elem.content.count > 1 && !child.is_a?(Text)
        if child.is_a?(Text)
           @valid_output << "Currently evaluating a Text :\n-Text -> Must contains a simple string\nText content is OK"
        end
        return false unless _is_recursive_valid?(child)
    end

    true
  end

  def to_s
    @elem.to_s
  end
end

if $PROGRAM_NAME == __FILE__
  # Example from subject
  toto = Html.new([
    Head.new([Title.new(Text.new("Hello ground!"))]),
    Body.new([
      H1.new(Text.new("Oh no, not again!")),
      Img.new([], {'src': Text.new('http://i.imgur.com/pfp3T.jpg')})
    ])
  ])
  test = Page.new(toto)
  test.is_valid?
end

File: Advanced Core/Django_RoR_Symfony--Piscines/Piscine-Django/2. Django - 0 - Oob/ex06/Page.py
from elem import Elem, Text
from elements import (Html, Head, Body, Title, Meta, Img, Table, Th, Tr, Td, 
                      Ul, Ol, Li, H1, H2, P, Div, Span, Hr, Br)

class Page:
    """
    Represents an HTML page with structural validation according to
    specific Django Piscine OOB rules.
    """
    def __init__(self, elem):
        """Initializes the Page with a root Elem instance."""
        if not isinstance(elem, Elem):
             raise TypeError("Page must be initialized with an Elem instance.")
        self.elem = elem

    def is_valid(self):
        """Verifies if the stored Elem tree is structurally valid."""
        return self._is_recursive_valid(self.elem)

    def _is_recursive_valid(self, elem):
        """
        Recursively checks each node against the permitted tags and
        parent-child relationship rules.
        """
        # Rule 1: Node type check - only specific classes and Text are allowed
        allowed_types = (Html, Head, Body, Title, Meta, Img, Table, Th, Tr, Td, 
                         Ul, Ol, Li, H1, H2, P, Div, Span, Hr, Br, Text)
        if not isinstance(elem, allowed_types):
            return False

        # If it's Text, it's valid (leaf node)
        if isinstance(elem, Text):
            return True

        # Check content recursively before checking the current node's rules
        for child in elem.content:
            if not self._is_recursive_valid(child):
                return False

        # Rule-specific tag constraints
        tag = elem.tag
        content = elem.content
        
        if isinstance(elem, Html):
            # Rule: Html must contain exactly [Head, Body]
            if len(content) != 2: return False
            if not isinstance(content[0], Head) or not isinstance(content[1], Body):
                return False

        elif isinstance(elem, Head):
            # Rule: Head must contain exactly one Title
            if len(content) != 1 or not isinstance(content[0], Title):
                return False

        elif isinstance(elem, (Body, Div)):
            # Rule: Body and Div can only contain layout/text elements
            body_allowed = (H1, H2, Div, Table, Ul, Ol, Span, Text)
            if not all(isinstance(c, body_allowed) for c in content):
                return False

        elif isinstance(elem, (Title, H1, H2, Li, Th, Td)):
            # Rule: These tags must contain exactly one Text node
            if len(content) != 1 or not isinstance(content[0], Text):
                return False

        elif isinstance(elem, P):
            # Rule: P tags can only contain Text nodes
            if not all(isinstance(c, Text) for c in content):
                return False

        elif isinstance(elem, Span):
            # Rule: Span can contain Text or P tags
            if not all(isinstance(c, (Text, P)) for c in content):
                return False

        elif isinstance(elem, (Ul, Ol)):
            # Rule: Lists must contain at least one Li and only Li tags
            if len(content) < 1 or not all(isinstance(c, Li) for c in content):
                return False

        elif isinstance(elem, Tr):
            # Rule: Tr must contain at least one cell; cells must be all Th or all Td
            if len(content) < 1: return False
            first_type = type(content[0])
            if first_type not in (Th, Td): return False
            if not all(isinstance(c, first_type) for c in content):
                return False

        elif isinstance(elem, Table):
            # Rule: Table can only contain Tr tags
            if not all(isinstance(c, Tr) for c in content):
                return False

        return True

    def __str__(self):
        res = ""
        if isinstance(self.elem, Html):
            res += "<!DOCTYPE html>\n"
        res += str(self.elem)
        return res

    def write_to_file(self, filename):
        with open(filename, "w") as f:
            f.write(str(self))

if __name__ == '__main__':
    # Test valid page
    valid_html = Html([
        Head(Title(Text('Hello World'))),
        Body([
            H1(Text('Title')),
            Div([
                P(Text('Para')),
                Span([Text('Span'), P(Text('Nested P'))])
            ])
        ])
    ])
    page = Page(valid_html)
    print(f"Valid page is_valid: {page.is_valid()}")
    
    # Test invalid page (Span containing Div)
    invalid_html = Html([
        Head(Title(Text('Fail'))),
        Body(Div(Span(Div(Text('Broken')))))
    ])
    page_fail = Page(invalid_html)
    print(f"Invalid page is_valid: {page_fail.is_valid()}")
